client/trino-cli/src/main/java/io/trino/cli/ClientOptions.java

```diff
@@ -22,35 +22,88 @@
 import io.airlift.units.Duration;
 import io.trino.client.ClientSession;
 import io.trino.client.auth.external.ExternalRedirectStrategy;
+import io.trino.client.uri.PropertyName;
+import io.trino.client.uri.RestrictedPropertyException;
+import io.trino.client.uri.TrinoUri;
 import okhttp3.logging.HttpLoggingInterceptor;
 import org.jline.reader.LineReader;
+import org.jline.reader.LineReaderBuilder;
 
 import java.net.URI;
 import java.net.URISyntaxException;
+import java.sql.SQLException;
 import java.time.ZoneId;
 import java.util.ArrayList;
+import java.util.Collections;
 import java.util.List;
 import java.util.Locale;
 import java.util.Map;
 import java.util.Objects;
 import java.util.Optional;
 import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
 
 import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
 import static com.google.common.base.Strings.nullToEmpty;
+import static io.trino.cli.TerminalUtils.getTerminal;
 import static io.trino.client.KerberosUtil.defaultCredentialCachePath;
+import static java.lang.String.format;
 import static java.util.Locale.ENGLISH;
 import static java.util.Objects.requireNonNull;
 import static picocli.CommandLine.Option;
+import static picocli.CommandLine.Parameters;
 
 public class ClientOptions
 {
     private static final Splitter NAME_VALUE_SPLITTER = Splitter.on('=').limit(2);
     private static final CharMatcher PRINTABLE_ASCII = CharMatcher.inRange((char) 0x21, (char) 0x7E); // spaces are not allowed
     private static final String DEFAULT_VALUE = "(default: ${DEFAULT-VALUE})";
-
-    @Option(names = "--server", paramLabel = "<server>", defaultValue = "localhost:8080", description = "Trino server location " + DEFAULT_VALUE)
-    public String server;
+    private static final String SERVER_DEFAULT = "localhost:8080";
+    private static final String SOURCE_DEFAULT = "trino-cli";
+
+    // some entries have different semantics, but this mapping is only used to disallow setting both as a CLI option and URL parameter
+    static final Map<String, PropertyName> optionsToProperties = ImmutableMap.<String, PropertyName>builder()
+            .put("--krb5-service-principal-pattern", PropertyName.KERBEROS_SERVICE_PRINCIPAL_PATTERN)
```

Comment:
can we add these bits to the `PropertyName` itself?

```
PropertyName(final String key, final String commandLineKey, boolean allowedInUri)
```

Reply:
This is a map of cli options to property names. Since the `PropertyName` class is defined in `trino-client` it shouldn't include cli options. When a user specifies both a URL and an option, we only allow the URL to contain properties that are not set as options, to avoid silently overriding them, which can be confusing. So all properties are allowed by default, unless some options are set by the user.

Reply:
Makes sense but how to ensure that those two lists are in-sync (CLI and PropertyName-s)?

Reply:
Let's introduce an annotation applied to the CLI Options with the PropertyName mapping

Reply:
I pulled in your suggestion